### PR TITLE
add locked_stake module

### PR DESCRIPTION
**Pull request by emmazzz**
## Description 

Added a package that allows locking sui tokens and stake objects up, and performing stake/unstake operations while the assets are locked.

## Test Plan 

Added tests.

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: examples/move/locked_stake/sources/locked_stake.move [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    /// Unstake the stake object with `staked_sui_id` and puts the resulting principal
+    /// and rewards back into the locked sui balance.
+    /// Returns the amount of SUI unstaked, including both principal and rewards.
+    /// Aborts if no stake exists with the given id.
+    public fun unstake(
+        ls: &mut LockedStake,
+        sui_system: &mut SuiSystemState,
+        staked_sui_id: ID,
+        ctx: &mut TxContext
+    ) : u64 {
```
Comment (sblackshear) [inline examples/move/locked_stake/sources/locked_stake.move:89]:
```suggestion
    ): u64 {
```

File: examples/move/locked_stake/sources/locked_stake.move [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    public fun sui_balance(ls: &LockedStake): u64 {
+        balance::value(&ls.sui)
+    }
+
+    public fun locked_until_epoch(ls: &LockedStake): u64 {
+        epoch_time_lock::epoch(&ls.locked_until_epoch)
+    }
+
+    // TODO: possibly add some scenarios like switching stake, creating a new LockedStake and transferring
+    // it to the sender, etc. But these can also be done as PTBs.
```
Comment (sblackshear) [inline examples/move/locked_stake/sources/locked_stake.move:113]: Ah, that is cool--if you do stake and unstake in the same PTB, there's no delay in performing the switch?

**Review by sblackshear — APPROVED** [review]
This looks great--thanks Emma!

File: examples/move/locked_stake/sources/locked_stake.move (hunk 2 above)
Comment (emmazzz) [reply examples/move/locked_stake/sources/locked_stake.move:113]: Right you lose one epoch of rewards but there's no delay.

**Pull request merged**